### ISSUE TITLE
The customized neuron has no output pulse/自定义神经元没有输出脉冲

### ISSUE THREAD

**Issue by jamieyang66** [fangwei123456/spikingjelly#324]
我尝试自定义一个神经元，功能是只激活一次，也就是TTFS神经元。我采用了类似https://spikingjelly.readthedocs.io/zh_CN/latest/activation_based/neuron.html 中的代码来实现这个神经元，代码如下：

    import torch
    from spikingjelly.activation_based import neuron
    import torch.nn as nn
    from spikingjelly.activation_based import surrogate
    from typing import Callable

    class Node(neuron.BaseNode):
        def __init__(self,  v_threshold: float = 1.,
                     v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                     detach_reset: bool = False, step_mode='s', backend='torch', store_v_seq: bool = False,f:float=0.):

            super().__init__(v_threshold, v_reset, surrogate_function, detach_reset, step_mode, backend, store_v_seq)
            self.f=f

        def neuronal_charge(self, x: torch.Tensor):
            self.v = self.v + (1-self.f)*x

        def neuronal_reset(self, spike):
            self.f=self.f+float(spike)

            if self.detach_reset:
                spike_d = spike.detach()
            else:
                spike_d = spike

            if self.v_reset is None:
                # soft reset
                self.v = self.jit_soft_reset(self.v, spike_d, self.v_threshold)

            else:
                # hard reset
                self.v = self.jit_hard_reset(self.v, spike_d, self.v_reset)

然后我用如下代码来进行验证：

    layer = Node()

    T = 4
    N = 1
    x_seq = torch.rand([T, N])
    print(f'x_seq={x_seq}')

    for t in range(T):
        yt = layer(x_seq[t])
        print(f'layer.v[{t}]={layer.v}')

    layer.reset()
    layer.step_mode = 'm'
    y_seq = layer(x_seq)
    print(f'y_seq={y_seq}')
    layer.reset()

发现虽然电压有变化，但是并没有脉冲输出：

    x_seq=tensor([[0.3026],
            [0.9712],
            [0.3170],
            [0.7839]])
    layer.v[0]=tensor([0.3026])
    layer.v[1]=tensor([0.])
    layer.v[2]=tensor([0.])
    layer.v[3]=tensor([0.])
    y_seq=tensor([[0.],
            [0.],
            [0.],
            [0.]])

但是如果单独使用代码

    layer = Node()

    T = 4
    N = 1
    x_seq = torch.rand([T, N])
    print(f'x_seq={x_seq}')

    layer.reset()
    layer.step_mode = 'm'
    y_seq = layer(x_seq)
    print(f'y_seq={y_seq}')
    layer.reset()
的话，又是有输出的：

    x_seq=tensor([[0.3192],
            [0.3813],
            [0.9214],
            [0.3755]])
    y_seq=tensor([[0.],
            [0.],
            [1.],
            [0.]])

因此我怀疑是reset函数并没有把self.f 归到0，不知道该如何解决

**Comment by fangwei123456** [comment]
如果 `self.f` 是状态，则不能简单的赋值初始化，而是用“注册记忆”函数，例如：

https://github.com/fangwei123456/spikingjelly/blob/6e537940513b5e5c94ff1f08a21c89f9bb386456/spikingjelly/activation_based/neuron.py#L100

"记忆"会在调用reset函数时进行重置：

https://github.com/fangwei123456/spikingjelly/blob/6e537940513b5e5c94ff1f08a21c89f9bb386456/spikingjelly/activation_based/base.py#L324

**Comment by jamieyang66** [comment]
改写成了这样，结果对了，不知道写法是否正确？

    class Node(neuron.BaseNode):
        def __init__(self,  v_threshold: float = 1.,
                     v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                     detach_reset: bool = False, step_mode='s', backend='torch', store_v_seq: bool = False):

            super().__init__(v_threshold, v_reset, surrogate_function, detach_reset, step_mode, backend, store_v_seq)
            self.register_memory('f', 0.)

        def neuronal_charge(self, x: torch.Tensor):
            self.v = self.v + (1-self.f)*x

        def neuronal_reset(self, spike):
            self.f=self.f+float(spike)

            if self.detach_reset:
                spike_d = spike.detach()
            else:
                spike_d = spike

            if self.v_reset is None:
                # soft reset
                self.v = self.jit_soft_reset(self.v, spike_d, self.v_threshold)

            else:
                # hard reset
                self.v = self.jit_hard_reset(self.v, spike_d, self.v_reset)

**Comment by fangwei123456** [comment]
这样是正确的，可以被重置。

**Comment by jamieyang66** [comment]
谢谢！
